cypress/integration/unit_tests/polygon_draw_test.js

```diff
@@ -205,208 +263,466 @@ describe('Unit test for ShapeData', () => {
-      expect(ids).to.eql(['my_id']);
-      expect(records).to.eql([{
-        calculatedData: calculatedData,
-        geometry: polygonGeometry,
-        notes: 'new notes',
-      }]);
-      expect(underTest.id).to.eql('my_id');
-      expect(StoredShapeData.pendingWriteCount).to.eql(0);
+    let popupPendingCalculationSpy;
+    let popupCalculatedDataSpy;
+    let eeSpy;
+    let firestoreSpy;
+    drawPolygonAndSetUpSpies().then((spyResult) => ({
+                                      popupPendingCalculationSpy,
+                                      popupCalculatedDataSpy,
+                                      eeSpy,
+                                      firestoreSpy,
+                                    } = spyResult));
+    cy.get('.notes').type(notes);
+    pressButtonAndWaitForPromise('save').then(() => {
+      expect(popupPendingCalculationSpy).to.not.be.called;
+      expect(popupCalculatedDataSpy).to.not.be.called;
+      expect(eeSpy).to.not.be.called;
+      expect(firestoreSpy).to.be.calledOnce;
+      assertOnFirestoreAndPopup(path, withNotes(notes));
     });
   });
 
+  // This test also incidentally tests that we correctly update Firestore on a
+  // dragged polygon.
+  it('Dragged polygon triggers recalculation', () => {
+    // Clone path and edit.
+    const newPath = JSON.parse(JSON.stringify(path));
+    newPath[0].lng = 0.5;
+    let popupPendingCalculationSpy;
+    let popupCalculatedDataSpy;
+    let firestoreSpy;
+    let eeSpy;
+    drawPolygonAndSetUpSpies().then((spyResult) => {
+      ({
+        popupPendingCalculationSpy,
+        popupCalculatedDataSpy,
+        firestoreSpy,
+        eeSpy,
+      } = spyResult);
+      getFirstFeature().setPath(newPath);
+    });
+    pressButtonAndWaitForPromise('save').then(() => {
+      expect(popupPendingCalculationSpy).to.be.calledOnce;
+      expect(popupCalculatedDataSpy).to.be.calledOnce;
+      expect(eeSpy).to.be.calledOnce;
+      expect(firestoreSpy).to.be.calledOnce;
+    });
+    assertOnFirestoreAndPopup(newPath, defaultData);
+  });
+
+  /**
+   * Draws the default polygon and clicks on it. Then returns spies to observe
+   * calculated-data-related calls made to the polygon's popup, and {@link
+   * ee#List} and
+   * {@link userShapes#doc} calls. Because EE is finicky about spying, the
+   * `eeSpy` returned is not actually spying on the real {@link ee#List} but
+   * rather on a dummy object that shadows it, and that we call with the same
+   * arguments as the real {@link ee#List}.
+   * @typedef {sinon.SinonSpy|sinon.SinonStub} Spy
+   * @return {Cypress.Chainable<{popupPendingCalculationSpy: Spy,
+   *     popupCalculatedDataSpy: Spy, eeSpy: Spy, firestoreSpy: Spy}>}
+   */
+  function drawPolygonAndSetUpSpies() {
+    let popupPendingCalculationSpy;
+    let popupCalculatedDataSpy;
+    let firestoreSpy;
+    const dummyObjectForSpyAssertions = {method: () => {}};
+    const eeSpy = cy.spy(dummyObjectForSpyAssertions, 'method');
```

Comment:
nit: can we call this something more description than method, like forSpying? Idk that's not great either but